Support third-party watchtowers in persistence pipeline

alecchendev opened the pull request:
Closes #1776. This PR aims to make it easier to form a justice/penalty transaction within the channel monitor persistence pipeline, which would help with building third-party watchtowers to support LDK-based nodes.

Still WIP, figuring out what is the best API for this. I currently have an implementation of `Persist` that can form a justice tx after exposing a few things on `ChannelMonitor` as an example of how I'd imagine a user being able to implement some sort of watchtower support. Open to feedback about how this should work.

Channel monitor persistence seems like the best place to have watchtower support so that we can have control over the channel pausing/resuming normal operation depending on the success/failure of persisting at the watchtower level. For this reason my initial approach was to see what was available from persistence and work with that, although I'm going to explore more what's available from the signer to see if I can improve things.

To do
- [ ] Try just using the signer to form the justice tx to get a better idea of what's there, which will likely help inform this API further
- [ ] Explore
- [ ]  If the reserve value is above the dust limit, the first commitment will have multiple outputs (right now I fixed the output index to 0), should account for this
- [ ]  Should test justice tx for commitments that are not just the first one
- [ ] `test_restored_packages_retry` is failing - has a hardcoded serialized channel monitor so anytime I add/change fields I'll have to change

alecchendev commented on lightning/src/util/test_utils.rs:
```diff
@@ -264,6 +265,157 @@ impl<'a> chain::Watch<EnforcingSigner> for TestChainMonitor<'a> {
+}
+
+impl<Signer: sign::WriteableEcdsaChannelSigner> chainmonitor::Persist<Signer> for WatchtowerPersister {
+	fn persist_new_channel(&self, funding_txo: OutPoint, data: &channelmonitor::ChannelMonitor<Signer>, id: MonitorUpdateId) -> chain::ChannelMonitorUpdateStatus {
+		assert!(self.channel_watchtower_state.lock().unwrap()
+			.insert(funding_txo, HashMap::new()).is_none());
+
+		// No ChannelMonitorUpdateStep::LatestCounterpartyCommitmentTxInfo is persisted on channel creation
+		// so we get the same info and update the watchtower
+		let commitment_number = (1u64 << 48) - 1;
+		let output_idx = 0;
+		let commitment_txid = data.get_current_counterparty_commitment_txid().expect("Should have at least one counterparty commitment tx");
+		let value = data.initial_to_counterparty_value();
```
Something important I've learned so far about my current approach: the very first counterparty commitment transaction when we open a channel is provided to the `ChannelMonitor` upon its creation, which means we don't get a `ChannelMonitorUpdateStep::LatestCounterpartyCommitmentTxInfo` in the `Persist` implementation for the first commitment tx, which is needed to form the justice tx. Right now I expose some extra stuff just for this case in a fairly hacky way. It may be worth trying to refactor to allow this initial update to go through persistence if it's thought that this sort of persistence API should be better supported, but generally this is making me lean towards exploring using what's already offered to the signer.

alecchendev commented:
Rewrote a bunch of this with a new approach which I think is better now. Still figuring out exactly the right API. I tried to elaborate in the commit messages on my thought process so far.

Also I still need to figure out how to best accommodate the very first monitor update with the counterparty's commitment, whether that be storing some state in the ChannelMonitor that gets cleaned up, or possibly trying to refactor to push the update through persistence, or something else.

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -506,6 +506,7 @@ pub(crate) enum ChannelMonitorUpdateStep {
 		htlc_outputs: Vec<(HTLCOutputInCommitment, Option<Box<HTLCSource>>)>,
 		commitment_number: u64,
 		their_per_commitment_point: PublicKey,
+		non_htlc_outputs: Vec<TxOut>,
```
I've reduced this down to mainly the to_local and to_remote values in sats, which from what I've found are the minimum amount of data needed to be able to reconstruct the commitment tx/find the revoked output data

alecchendev commented on lightning/src/util/test_utils.rs:
```diff
@@ -264,6 +265,157 @@ impl<'a> chain::Watch<EnforcingSigner> for TestChainMonitor<'a> {
+		if let Some(up) = update {
+			for step in up.updates.iter() {
+				match step {
+					channelmonitor::ChannelMonitorUpdateStep::LatestCounterpartyCommitmentTXInfo { commitment_txid, htlc_outputs: _, commitment_number, their_per_commitment_point, non_htlc_outputs } => {
```
> Note that these are not actually public,

Ah my bad!

> instead exposing methods on `ChannelMonitorUpdate` which allow users access to the set of new commitment transactions which could be broadcast by us or our counterparty and then letting them ask us to pre-sign claims of our counterparty's outputs.

Doing this exactly as described is a bit difficult, but going off of this I think I've made some progress in making this API more intuitive

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2591,6 +2634,180 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
+		let their_per_commitment_point = PublicKey::from_secret_key(&self.onchain_tx_handler.secp_ctx, &per_commitment_key);
+		let revokeable_redeemscript = self.revokeable_redeemscript(&their_per_commitment_point);
+
+		let sig = self.onchain_tx_handler.signer.sign_justice_revoked_output(&justice_tx, input_idx, value, &per_commitment_key, &self.onchain_tx_handler.secp_ctx)?;
```
Between the two approaches mentioned “Get counterparty commitment txs from monitor update” and “Get revoked output data from update, build/sign separately”, I think one of the determinant factor to pick up which one might be the most advantageous is the flexibility preserved for the watchtower.

Namely, they are few characteristics than a watchtower signing policy can enforce:
- the amount of fees paid by the `justice_tx`
- the destination `scriptPubkey`
- the claim aggregation of one or more `RevokeableOutput`

In term of a potential watchtower signing policy, from my understanding the two approaches are equivalent:
- with approach 1, as we’re calling into `OnchainTxHandler`’s signer, a signing policy can be implemented behind `EcdsaChannelSigner::sign_justice_revoked_output`
- with approach 2, `build_justice_tx` is a pub method, the caller can verify the checks on the `Transaction` and then give it back to `sign_justice_tx`.

As we have already refactored our signing interface to let custom signing policy be implemented there, this is unclear to me what are the other trade-off between both approaches ? Do we wanna the transaction broadcast to be managed by the own transaction-relay module of the watchtower ? Or rather extend `OnchainTxHandler` to have `justice_tx` re-scheduled by the watchtower.

ariard replied:
> (in case in the future a watchtower module wanted to do anything beyond the most basic justice tx, e.g. add an output to the justice tx like mentioned [here](https://github.com/sr-gi/bolt13/blob/master/13-watchtowers.md#payment-modes) or claiming revoked HTLC outputs).

Yes adding an output as a “service bounty” for the outsourced watchtower makes a lot of sense (or fee-bumping purpose if you wanna do CPFP). Note, I would favor to expose the full commitment transaction data that way the watchtower can build claim on the revoked HTLC outputs, without waiting counterparty confirmation of second-stage HTLC transaction (which might never happen!). The anchor output could be swept too by the watchtower as recommended by BOLT3 after the CSV 16: https://github.com/lightning/bolts/blob/master/03-transactions.md#to_local_anchor-and-to_remote_anchor-output-option_anchors (not punishment per se though as you have the infrastructure laid out).

> As I've had the basic "outsourced" watchtower in mind for this, I imagined that this would expose the necessary information to build, sign, and encrypt the justice tx, and send it to the watchtower server which would take care of broadcasting the transaction separately.

“Outsourced” watchtower sounds good. There is one design option moving forward, making `OnchainTxHandler` an interface to which transaction issuer (watchtower, dual-funding rbfing) can subscribe broadcasting transaction. A robust transaction rebroadcast and fee-bumping module is easier said than done :) (especially with all transaction relay policy maintenance).

ariard reviewed (commented):
> Still WIP, figuring out what is the best API for this. I currently have an implementation of
> Persist that can form a justice tx after exposing a few things on ChannelMonitor as an
> example of how I'd imagine a user being able to implement some sort of watchtower
> support. Open to feedback about how this should work.

I think this would benefit the review of this PR if we have a more detailed flow of how do you assume third-party watchtowers flows to work.

If we assume “local” watchtowers or “outsourced” ones (i.e run by an external entity where the pre-signed justice transactions are potentially encrypted thanks to BOLT3’s obscured commitment number). And if the flow should work for basic watchtower (justice transaction-only) or monitor replica (justice txn + time-sensitive ones like HTLC claims), see our GLOSSARY here.

 There is a historical BOLT13 on how watchtower might work: https://github.com/sr-gi/bolt13/blob/master/13-watchtowers.md, however to the best of my knowledge there is no documentation on watchtower architecture (which might be helpful not only for cross-implementation compatibility but also guarantees compatibility between versions of LDK).

alecchendev reviewed (commented):
> I think this would benefit the review of this PR if we have a more detailed flow of how do you assume third-party watchtowers flows to work.

Ah thanks, I'll add that to the PR description. I've been mainly designing this with the "basic" (justice tx-only) "outsourced" (pre-signed, encrypted justice tx sent to an external entity) watchtower in mind.

sr-gi commented:
I've tried this downstream on `ldk-sample`. The workflow, from an API standpoint, is what I'd expect.

If I've got this right, for a client to implement the functionality something like [`WatchtowerPersister`](https://github.com/lightningdevkit/rust-lightning/pull/2337/files#diff-f68cd0fdd831a4f8bfafb3d25004fd2315ce5def88ffb62b667d658dab4bfd0dR271) would need to be implemented so the state can be backed up to the remote tower. In my test, I have just logged the methods on the persister and tried this on regtest against a CLN node. In a prod setup, it would make more sense to have a queue that the client process can subscribe to, rather than a map, but I guess that's up to the implementer of the persister to pick whatever they find more useful.

Here's the patched version of `ldk-sample` building on top of @tnull's 0.0.116 update WIP and pulling from a custom `rust-lightning` with this patch: https://github.com/sr-gi/ldk-sample/tree/ldk-towers

alecchendev commented:
Thanks for testing it out @sr-gi!

> If I've got this right, for a client to implement the functionality something like [`WatchtowerPersister`](https://github.com/lightningdevkit/rust-lightning/pull/2337/files#diff-f68cd0fdd831a4f8bfafb3d25004fd2315ce5def88ffb62b667d658dab4bfd0dR271) would need to be implemented so the state can be backed up to the remote tower.

Yep, your patch looks like how I was imagining it to be used

> In a prod setup, it would make more sense to have a queue that the client process can subscribe to, rather than a map, but I guess that's up to the implementer of the persister to pick whatever they find more useful.

Makes sense, yea I just used a map off the top of my head for testing :)

ariard commented:
Concept ACK, I’ll favor approach 1 over approach 2 as exposing the full commitment data allows more flexibility for the outsourced tower service (e.g fee-bumping output).

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1322,6 +1340,31 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	}
+
+	/// TODO: docs
+	pub fn counterparty_commitment_txs_from_update(&self, update: &ChannelMonitorUpdate) -> Vec<CommitmentTransaction> {
```
Great idea!

alecchendev commented:
Rebased, cleaned things ups, added docs, and reworked some things according to all the feedback so far

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1376,6 +1412,29 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	}
+
+	/// Gets all of the counterparty commitment transactions provided by the given update. This
+	/// may be empty if the update doesn't include any new counterparty commitments.
```
Added significantly to all the added public monitor methods

alecchendev commented:
Just fixed a docs build error

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2686,6 +2738,17 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
 
+	pub(crate) fn initial_counterparty_commitment_tx(&mut self) -> Option<CommitmentTransaction> {
+		let (their_per_commitment_point, feerate_per_kw, to_broadcaster_value,
+			to_countersignatory_value) = self.initial_counterparty_commitment_info.take()?;
```
We probably don't need to, but I couldn't really think of another time when a user would want to call this besides on the first persist, and thought if we took here we wouldn't have to keep around and persist extra bytes (although it's not really much) - could see it either way, it might just be a simpler API to not have to think about whether it's been called before though?

alecchendev commented on lightning/src/ln/channel.rs:
```diff
@@ -2512,7 +2512,12 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 		                                          holder_commitment_tx, best_block, self.context.counterparty_node_id);
 
-		channel_monitor.provide_latest_counterparty_commitment_tx(counterparty_initial_bitcoin_tx.txid, Vec::new(), self.context.cur_counterparty_commitment_transaction_number, self.context.counterparty_cur_commitment_point.unwrap(), logger);
+		channel_monitor.provide_initial_counterparty_commitment_tx(
```
Hm, storing the whole commitment tx object is 488 bytes whereas the current is 52 bytes. Maybe in the grand scheme of a monitor it might not be thaat much, but I feel like that's quite a bit more to store just to save on a computation we only expect the user to call once?

alecchendev replied:
Oh I was just going off the `CommitmentTransaction` struct - if we stored the built `Transaction` it's less, but then we'd still probably want to rebuild the `CommitmentTransaction` struct when we provide it to the user?

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1384,6 +1384,23 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	/// justice transactions for any revoked commitment transactions. It's intended to be used
+	/// within an implementation of [`Persist::update_persisted_channel`], which is provided with
+	/// a monitor and an update. Note that this simply provides the counterparty commitment
+	/// transactions from an update, which may or may not have been revoked yet.
```
Just removed it since I don't think it's really adding much

alecchendev commented:
Squashed because there were a million fixups, added one fixup for higher level docs to make it easier to find that you can assemble all these added methods to build a watchtower

ariard commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -1607,6 +1613,62 @@ impl CommitmentTransaction {
 
+	/// Returns the index of the revokeable output in the built transaction, if any exists.
+	/// There are two cases where this may return `None`:
+	/// - The balance of the revokeable output is below the dust limit (only found on commitments
```
the HTLC outputs on a commitment transaction as also revokable, could be called the `to_local` output as it's only balance output you can revoke when you own the commitment. beyond, I wonder if the first case isn't more when the balance of the revokable output is between dust limit and channel reserve, if it's below the dust limit it should be removed according to bolt3 commitment transaction construction and our `build_commitment_transaction`, I think

alecchendev replied:
> the HTLC outputs on a commitment transaction as also revokable, could be called the to_local output as it's only balance output you can revoke when you own the commitment.

Good point, added some clarification that it's the revokeable output sending funds to the broadcaster

> beyond, I wonder if the first case isn't more when the balance of the revokable output is between dust limit and channel reserve, if it's below the dust limit it should be removed according to bolt3 commitment transaction construction and our build_commitment_transaction, I think

Our channel reserve is required to be above the dust limit so I think this is fine?

ariard replied:
> Our channel reserve is required to be above the dust limit so I think this is fine?

It’s correct our channel is required to be above the dust limit (we check against `MIN_CHAN_DUST_LIMIT_SATOSHIS` in `OpenChannel` reception) though I was pointing you might have a `to_local_output` amount between dust_limit_satoshis and channel_reserve during early channel lifetime (e.g you received one HTLC above `dust_limit_satoshis` though under the 1% of the channel capacity, our default value iirc), though once channel_reserve is met it’s cannot go under. Comment was a bit confusing to me, nevermind to change it.

ariard commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -1607,6 +1613,62 @@ impl CommitmentTransaction {
+	///
+	/// This method will error if the given fee rate results in a fee greater than the value
+	/// of the output being spent.
+	pub fn build_justice_tx(&self, output_idx: u32, feerate_per_kw: u64, lock_time: u32, destination_script: Script)
```
The lock time could be removed, or advice could be given if anti-fee snipping technique is adopted by tweaking the lock_time, see: https://github.com/bitcoin/bitcoin/pull/2340

alecchendev replied:
Hmm, since this is expected to be a pre-signed tx that will be broadcast at an unknown point in the future, is it possible to meaningfully influence fee-sniping by tweaking the locktime?

alecchendev replied:
Done

ariard commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -1607,6 +1613,62 @@ impl CommitmentTransaction {
+	/// on this transaction to a destination script. Use [`Self::revokeable_output_index`] to get
+	/// the index of the output to spend to create a valid justice transaction. Fee estimation
+	/// assumes the transaction is spending a revoked output, accounting for the expected
+	/// witness data that will be added when signed.
```
can say the transaction is RBFable and a reasonable `feerate_per_kw` can be adopted for first sequence of broadcast

alecchendev replied:
Added a note about RBF, as well as that the user may build multiple copies at different feerates :)

ariard reviewed (commented):
code looks good, haven't look on the tests

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1376,6 +1412,60 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	/// This is provided so that watchtower clients in the persistence pipeline are able to build
+	/// justice transactions for each counterparty commitment upon each update. It's intended to be
+	/// used within an implementation of [`Persist::update_persisted_channel`], which is provided
+	/// with a monitor and an update. Once revoked, signing a justice transaction can be done using
```
Yea, the case I'm expecting is that the justice tx is signed in advance and then the broadcast by the watchtower server, but true the user could wait to till broadcast to sign in another setup. Are docs here fine then considering this since it doesn't specifically suggest either way?

alecchendev commented:
Rebased for CI conflict (just had to do `s/claimable_amount_satoshis/amount_satoshis` on one line near the end of the test)

alecchendev commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -1607,6 +1613,62 @@ impl CommitmentTransaction {
+	///
+	/// This method will error if the given fee rate results in a fee greater than the value
+	/// of the output being spent.
+	pub fn build_justice_tx(&self, output_idx: u32, feerate_per_kw: u64, lock_time: u32, destination_script: Script)
```
Went with `build_to_local_justice_tx` to stay consistent with how it's referred to in the spec. Also left `revokeable_output_index` public in case a user wants to build the tx themselves

alecchendev commented on lightning/src/ln/channel.rs:
```diff
@@ -5276,17 +5283,23 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 				commitment_number: self.context.cur_counterparty_commitment_transaction_number,
-				their_per_commitment_point: self.context.counterparty_cur_commitment_point.unwrap()
+				their_per_commitment_point: self.context.counterparty_cur_commitment_point.unwrap(),
+				feerate_per_kw: Some(self.context.feerate_per_kw),
```
Ah true, replaced with the feerate from the built `CommitmentTransaction`

alecchendev commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -2551,6 +2551,70 @@ fn revoked_output_claim() {
 }
 
+#[test]
+fn test_forming_justice_tx_from_monitor_updates() {
```
Oof yea, makes sense. I'd originally planned to come back and think about coverage more, and then forgot, sigh... Added debug asserts for the initial + following counterparty commitment reconstruction, which ended up catching that I hadn't filtered out dust htlc outputs, which I've added a commit for now.

After some runs of a simple `time cargo test` comparing this branch with main it doesn't seem to make a noticeable difference, not sure if you had more sophisticated benchmarking in mind for this

alecchendev commented:
Rebased for CI conflict, added commits to filter out dust HTLCs when reconstructing the commitment tx, added a basic unit test for the justice tx builder helper, addressed review feedback

alecchendev commented:
Squashed without further changes

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2651,6 +2778,107 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
+					to_broadcaster_value_sat: Some(to_broadcaster_value),
+					to_countersignatory_value_sat: Some(to_countersignatory_value) } => {
+
+					let dust_limit_satoshis = self.counterparty_dust_limit_satoshis?;
+					let channel_type = self.onchain_tx_handler.channel_type_features();
+					let nondust_htlcs = htlc_outputs.iter().filter_map(|htlc_and_source| {
+						let (htlc, _) = htlc_and_source;
+						// This logic to determine whether an HTLC is dust MUST exactly match
+						// the logic in ChannelContext::build_commitment_transaction
```
Wondering, why does the monitor care about our counterparty's dust HTLCs anyway since it doesn't need to watch them to claim? Another option could be to filter out dust HTLCs when we create the monitor update instead of here

alecchendev replied:
Oh ok gotcha

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2651,6 +2778,107 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
+					let channel_type = self.onchain_tx_handler.channel_type_features();
+					let nondust_htlcs = htlc_outputs.iter().filter_map(|htlc_and_source| {
+						let (htlc, _) = htlc_and_source;
+						// This logic to determine whether an HTLC is dust MUST exactly match
+						// the logic in ChannelContext::build_commitment_transaction
```
O thanks

alecchendev replied:
Done

alecchendev commented:
Just simplified filtering the dust outputs and removed the commit adding `counterparty_dust_limit_satoshis`, squashed immediately

alecchendev commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1446,6 +1447,25 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	/// `Value` is the value of the output being spent by the input at `input_idx`, committed
+	/// in the BIP 143 signature.
+	///
+	/// This method will only succeed if this monitor has received the revocation secret for the
+	/// provided `commitment_number`. If a commitment number is provided that does not correspond
+	/// to the commitment transaction being revoked, this will return a signed transaction, but
+	/// the signature will not be valid.
```
yea hmm, I feel like they'd have to really go out of their way to spend the wrong output/commitment number since they should just be using our helper(s). I don't feel that strongly either way, do you feel like this will meaningfully catch user errors?

alecchendev commented on lightning/src/util/test_utils.rs:
```diff
@@ -269,6 +271,108 @@ impl<'a> chain::Watch<EnforcingSigner> for TestChainMonitor<'a> {
+	) -> chain::ChannelMonitorUpdateStatus {
+		let res = self.persister.update_persisted_channel(funding_txo, update, data, update_id);
+
+		if let Some(update) = update {
```
Sure, will try and get to this over the next couple of days

alecchendev replied:
Sorry for the delay here, getting to this now. Were you thinking to expose basically a cleaned up version of this test util? I'm wondering who would this be for, since I imagine someone using this interface would want to return a different `ChannelMonitorUpdateStatus` depending on however they're using these signed justice txs (e.g. sending to remote watchtower/other server).

Hmm although... I could extend this util to take some user-implemented interface that takes a signed justice tx, does something, and returns a `ChannelMonitorUpdateStatus`, so our util handles the tracking and the user can focus on handling what to do with the tx? Would that make sense?

alecchendev replied:
Ah ok that makes sense!

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1376,6 +1376,29 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitor<Signer> {
+	///
+	/// It is expected that a watchtower client may use this method to retrieve the latest counterparty
+	/// commitment transaction(s), and then hold the necessary data until a later update in which
+	/// the monitor has been updated with the corresponding revocation data, at which point the
```
nit: s/revocation data/`per_commitment_secret`/g

ariard replied:
More confusing to have loosely names imho.

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2657,6 +2684,55 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
+	fn build_counterparty_commitment_tx(
+		&self, commitment_number: u64, their_per_commitment_point: &PublicKey,
+		to_broadcaster_value: u64, to_countersignatory_value: u64, feerate_per_kw: u32,
+		mut htlc_outputs: Vec<(HTLCOutputInCommitment, Option<Box<HTLCSource>>)>
```
nit: can say “nondust_htlcs` here, never included in an on-chain commitment tx.

ariard commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -1724,6 +1731,69 @@ impl<'a> TrustedCommitmentTransaction<'a> {
+	/// The built transaction will allow fee bumping with RBF, and this method takes
+	/// `feerate_per_kw` as an input such that multiple copies of a justice transaction at different
+	/// fee rates may be built.
+	pub fn build_to_local_justice_tx(&self, feerate_per_kw: u64, destination_script: Script)
```
This might sounds obvious though the `destination_script` and all the descendant scripts should be only controlled by the local lightning node, to avoid all kind of boring pinnings. If you like more doc.

ariard reviewed (approved):
Code Review ACK a21f2428, haven’t look deep into the tests.

alecchendev commented:
Oops forgot to say I rebased because of conflicts

alecchendev commented:
Squashed with no further changes
